### DISCUSSION TITLE
Potential bug in `getLogs`

### DISCUSSION THREAD

**Discussion by benesjan** [wevm/viem#366] [Question] [answered]
Hello,
I am fetching logs and I get a log from an older block than `fromBlock` param. I am testing the code against Anvil. Is this a bug or is this for whatever reason an expected behavior? Thank you

<img width="1785" alt="image" src="https://user-images.githubusercontent.com/13470840/231520834-2912e792-8c84-45d4-a9ab-5b0f92ed1b8f.png">

**Comment by jxom** [comment]
Interesting! We definitely do not modify the block number sent over RPC. Could be worth raising this as an issue in Foundry. 

https://github.com/foundry-rs/foundry

**Reply by benesjan** [reply]
When, I do the request with curl it does return only 1 event from block 6 and not from block 4. But when I isolate just the specific call with viem it also behaves correctly and returns one log from block 6.

Is there a way to make viem print out the JSON_RPC requests it's doing? Thank you

**Reply by jxom** [reply]
Can you just modify the `viem/esm` (or `viem/cjs`) module in `node_modules` and add a `console.log`?

**Reply by benesjan** [reply]
I did as you say and I think it's a caching issue. This is the modified code:
<img width="774" alt="image" src="https://user-images.githubusercontent.com/13470840/231698058-b1447b73-e04f-4725-ba17-5f0fa32fcd1f.png">

and this is the output in console:
```
  console.log
    request {
      method: 'eth_getLogs',
      params: [
        {
          address: '0xe7f1725E7734CE288F8367e1Bb143E90bb3F0512',
          topics: [Array],
          fromBlock: '0x0',
          toBlock: undefined
        }
      ]
    } result [
      {
        address: '0xe7f1725e7734ce288f8367e1bb143e90bb3f0512',
        topics: [
          '0xb72b85ef2a843244f5ea1955248b0ac363732d2e6a98cc3641084dd5718ad8b5',
          '0x0000000000000000000000000000000000000000000000000000000000000001',
          '0x000000000000000000000000f39fd6e51aad88f6f4ce6ab8827279cfffb92266'
        ],
        data: '0x00000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000154000001508ec32008eb76ebaed4df38fe8c6b807ccd60aeaffdfea6326454cab6e92eaecbb23ec1bdd1362e79ca51dcbdd77ea1a48489609b7aff189d439eca096f957d17c3eacc0c8013db582aea22c429c90aa8820f39da29611b5c9b08f50a633ae484e3fa42390f614661b1ac7b3828575e78b0199516e5237f235b11532fc0b40fcfa7734af2486589162d4950d86d60d30c9af1e844a1d24650021c0a364b070f6669e52e2799bb05583fdfa5a61c49f9eb9561450955a2203faea4bd65d33228ed52ffc312cfd5375ad142ffcf1046134071364dd3ec012d2574875fa5d761cdcdf54d7ef15725acd971ac4be78e9be9f8c7e6e3c3506efa333eb3c43cd800a76efba9c033137ef4a77eb4d87bf03815901027d4634189a33e6672548d4d074d479b36151a26517ce56b8811c6fab0e75f0efb0ba753fe1bb02f26dc41e3086935078e462de6a40f99c34a6b7da3ce52f7000000000000000000000000',
        blockHash: '0xb02bdc8c3f84300fa50b23b8a2717c1926fcdea156b14b5494e19e4079ce14cb',
        blockNumber: '0x4',
        transactionHash: '0xceb81834cc14106bc154816b83da9f2d7890db5d672edb3331eb3a740fcafcf3',
        transactionIndex: '0x0',
        logIndex: '0x0',
        transactionLogIndex: '0x0',
        removed: false
      }
    ]

      at getLogs (../../node_modules/viem/src/actions/public/getLogs.ts:125:14)

  console.log
    request {
      method: 'eth_getLogs',
      params: [
        {
          address: '0xe7f1725E7734CE288F8367e1Bb143E90bb3F0512',
          topics: [Array],
          fromBlock: '0x5',
          toBlock: undefined
        }
      ]
    } result [
      {
        address: '0xe7f1725e7734ce288f8367e1bb143e90bb3f0512',
        topics: [
          '0xb72b85ef2a843244f5ea1955248b0ac363732d2e6a98cc3641084dd5718ad8b5',
          '0x0000000000000000000000000000000000000000000000000000000000000001',
          '0x000000000000000000000000f39fd6e51aad88f6f4ce6ab8827279cfffb92266'
        ],
        data: '0x00000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000154000001508ec32008eb76ebaed4df38fe8c6b807ccd60aeaffdfea6326454cab6e92eaecbb23ec1bdd1362e79ca51dcbdd77ea1a48489609b7aff189d439eca096f957d17c3eacc0c8013db582aea22c429c90aa8820f39da29611b5c9b08f50a633ae484e3fa42390f614661b1ac7b3828575e78b0199516e5237f235b11532fc0b40fcfa7734af2486589162d4950d86d60d30c9af1e844a1d24650021c0a364b070f6669e52e2799bb05583fdfa5a61c49f9eb9561450955a2203faea4bd65d33228ed52ffc312cfd5375ad142ffcf1046134071364dd3ec012d2574875fa5d761cdcdf54d7ef15725acd971ac4be78e9be9f8c7e6e3c3506efa333eb3c43cd800a76efba9c033137ef4a77eb4d87bf03815901027d4634189a33e6672548d4d074d479b36151a26517ce56b8811c6fab0e75f0efb0ba753fe1bb02f26dc41e3086935078e462de6a40f99c34a6b7da3ce52f7000000000000000000000000',
        blockHash: '0xb02bdc8c3f84300fa50b23b8a2717c1926fcdea156b14b5494e19e4079ce14cb',
        blockNumber: '0x4',
        transactionHash: '0xceb81834cc14106bc154816b83da9f2d7890db5d672edb3331eb3a740fcafcf3',
        transactionIndex: '0x0',
        logIndex: '0x0',
        transactionLogIndex: '0x0',
        removed: false
      }
    ]
```

I guess the caching doesn't correctly take into account the `fromBlock` param and for this reason it returns incorrect response the second time. When I run the request in isolation either with curl or with viem it returns an event from **block 6**. So essentially when the caching is active I receive only the old event from block 4 and not the new one from block 6 on the second request.

 Can I somehow disable the caching? I would like to get it working.
 
 I am using the newest version of Viem (0.3.1).

**Reply by jxom** [reply]
There is no caching on `getLogs`. It simply forwards it to the RPC. 

Can you put a log here to confirm it’s not an RPC issue? https://github.com/wagmi-dev/viem/blob/main/src/utils/rpc.ts#L83

If it still returns the log, it might be an issue in viem; if it doesn’t, could be something happening on the RPC Provider.

Think you could provide me with a reproducible example so I can look into it further?

**Reply by fubhy** [reply]
Anvil has caching for fork requests. Maybe the issue is there?

**Reply by fubhy** [reply]
Are you using a fork? Can you try to run anvil with the `--no-storage-caching` option?

**Reply by benesjan** [reply]
I am not using a fork and the `--no-storage-caching` param is not available without fork.
<img width="420" alt="image" src="https://user-images.githubusercontent.com/13470840/231718197-2517f91f-fe2d-42dd-bc6d-1de80ffd4dac.png">

When I place the log in `rpc.js` I get the same incorrect output and when I copy the request body from the output and run it with curl I get the correct result. For this reason I would say it has to be some issue on the request side and not in Anvil. I tried explicitly setting  `no-cache` in fetch call but it didn't help. It's really weird. Will try to do the reproducible example.

<img width="2424" alt="image" src="https://user-images.githubusercontent.com/13470840/231728055-19d3cfb0-6381-4ed6-9915-698c69d409c2.png">

**Reply by benesjan** [reply]
I tried logging all http traffic with `sudo tcpflow -p -c -i lo0 port 9000` (started anvil on port 9000) and then when I pasted the exact request to curl again I got the incorrectly cached response directly from anvil. I guess I didn't catch this before because maybe the json was slightly different or something. Will talk to Foundry devs and close this if it will actually turn out to be anvil issue.
<img width="917" alt="image" src="https://user-images.githubusercontent.com/13470840/231755114-367c3337-ff64-4f7c-a601-c420ed9e7648.png">

**Reply by benesjan** [reply]
I didn't catch the anvil issue before because it seems that when I don't kill the test immediately after the bug occurs the test does further communication with anvil and it makes the bug not occur again. So I managed to only reproduce the bug with curl when I carefully watch the logs and kill the test fast.

**Comment by benesjan** [comment, marked as answer]
It was a bug in Anvil as per [this comment](https://github.com/foundry-rs/foundry/issues/4729#issuecomment-1507122562).
